block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -223,19 +228,39 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
 
         // Add latest messages to DB
         _ <- latestMessagesIndex.put(newLatestToAdd.toList)
+
+        // if block added as approved, record it as directly finalized.
+        _ <- blockMetadataIndex.recordFinalized(blockMetadata.blockHash, Set.empty).whenA(approved)
+
       } yield ()
     }
 
     lock.withPermit(
-      blockMetadataIndex.contains(block.blockHash).ifM(logAlreadyStored, doInsert) >> representation
+      blockMetadataIndex
+        .contains(block.blockHash)
+        .ifM(logAlreadyStored, doInsert) >> representation
     )
   }
 
   override def accessEquivocationsTracker[A](f: EquivocationsTracker[F] => F[A]): F[A] =
     lock.withPermit(f(KeyValueStoreEquivocationsTracker))
 
-  def addFinalizedBlockHash(blockHash: BlockHash): F[Unit] =
-    blockMetadataIndex.addFinalizedBlock(blockHash)
+  /** Record that some hash is directly finalized (detected by finalizer and becomes LFB). */
+  def recordDirectlyFinalized(
+      directlyFinalizedHash: BlockHash,
+      finalizationEffect: Set[BlockHash] => F[Unit]
+  ): F[Unit] =
+    for {
+      dag    <- representation
+      errMsg = s"Attempting to finalize nonexistent hash ${PrettyPrinter.buildString(directlyFinalizedHash)}."
+      _      <- dag.contains(directlyFinalizedHash).ifM(().pure, new Exception(errMsg).raiseError)
+      // all non finalized ancestors should be finalized as well (indirectly)
+      indirectlyFinalized <- dag.ancestors(directlyFinalizedHash, dag.isFinalized(_).not)
+      // invoke effects
+      _ <- finalizationEffect(indirectlyFinalized + directlyFinalizedHash)
+      // persist finalization
+      _ <- blockMetadataIndex.recordFinalized(directlyFinalizedHash, indirectlyFinalized)
+    } yield ()
```

Comment:
This is to dangerous to be without locking. I cannot say the use case right now where can fail but in parallel replay finalization can be called on overlapping blocks.

Reply:
But why this is a problem? Data put is the same

Reply:
https://github.com/rchain/rchain/pull/3459/commits/6b5d545bb0a07ec63bbdf28aafa1080cd48fbc98